---
title: "Pacuta Early Life History Temp TagSeq"
author: "Ariana S Huffmyer"
date: "2/18/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

Preliminary analysis of TagSeq dataset for Pocillopora acuta Moorea early life history temperature exposure.  


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

Load required libraries.  

```{r}
if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse') 
if ("genefilter" %in% rownames(installed.packages()) == 'FALSE') install.packages('genefilter') 
if ("DESeq2" %in% rownames(installed.packages()) == 'FALSE') install.packages('DESeq2') 
if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer') 
if ("WGCNA" %in% rownames(installed.packages()) == 'FALSE') install.packages('WGCNA') 
if ("flashClust" %in% rownames(installed.packages()) == 'FALSE') install.packages('flashClust') 
if ("gridExtra" %in% rownames(installed.packages()) == 'FALSE') install.packages('gridExtra') 
if ("ComplexHeatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('ComplexHeatmap') 
if ("goseq" %in% rownames(installed.packages()) == 'FALSE') install.packages('goseq') 
if ("dplyr" %in% rownames(installed.packages()) == 'FALSE') install.packages('dplyr') 
if ("clusterProfiler" %in% rownames(installed.packages()) == 'FALSE') install.packages('clusterProfiler') 
if ("pheatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('pheatmap') 
if ("magrittr" %in% rownames(installed.packages()) == 'FALSE') install.packages('magrittr') 
if ("vegan" %in% rownames(installed.packages()) == 'FALSE') install.packages('vegan') 

library("tidyverse")
library("genefilter")
library("DESeq2")
library("RColorBrewer")
library("WGCNA")
library("flashClust")
library("gridExtra")
library("ComplexHeatmap")
library("goseq")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("vegan")

```

# **Data input and filtering**  
 
Import the data files.  

```{r}
#load metadata sheet with sample name and developmental stage information
metadata <- read.csv("Pacu2021/Data/TagSeq/Sample.Info.csv", header = TRUE, sep = ",")
head(metadata)

#load gene count matrix generated from cluster computation
gcount <- as.data.frame(read.csv("Pacu2021/Output/TagSeq/Pacuta_gene_count_matrix.csv", row.names="gene_id"), colClasses = double)
head(gcount)
```

Check that there are no genes with 0 counts across all samples. 

```{r}
gcount<-gcount %>%
     mutate(Total = rowSums(.[, 1:18]))%>%
    filter(!Total==0)%>%
    select(!Total)
```

We started with 38,585 genes. About 18,000 genes that had 0's across all samples with 20,518 remaining.    

Conduct data filtering, this includes:  

*pOverA*: Specifying the minimum count for a proportion of samples for each gene. Here, we are using a pOverA of 0.05. This is because we have 18 samples with a minimum of n=2 samples per group. Therefore, we will accept genes that are present in 2/18 = 0.1 of the samples because we expect different expression by coral colony We are further setting the minimum count of genes to 10, such that 10% of the samples must have a gene count of >10 in order for the gene to remain in the data set.  

Filter in the package "genefilter". Pre-filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests.   

```{r}
filt <- filterfun(pOverA(0.1,10))

#create filter for the counts data
gfilt <- genefilter(gcount, filt)

#identify genes to keep by count filter
gkeep <- gcount[gfilt,]

#identify gene lists
gn.keep <- rownames(gkeep)

#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keep),])

#How many rows do we have before and after filtering?
nrow(gcount) #Before
nrow(gcount_filt) #After
```

Before filtering, we had approximately 20,518 genes. After filtering for pOverA, we have approximately 8,053 genes. This indicates that there were about 8,053 genes present in 10% of samples at <10 counts per gene.  

In order for the DESeq2 algorithms to work, the SampleIDs on the metadata file and count matrices have to match exactly and in the same order. The following R clump will check to make sure that these match. Should return TRUE.  
```{r}
#Checking that all row and column names match. Should return "TRUE"
all(rownames(metadata$sample_id) %in% colnames(gcount_filt))
all(rownames(metadata$sample_id) == colnames(gcount_filt)) 
```

Returns TRUE.  

# **Format metadata and count data columns**  

Change column names to be the numerical tube number. FIX THIS FOR OURS TO Pact##  
```{r}
result<-sub("_.*", "", colnames(gcount_filt)) #keep characters before _
#result2<-sub(".*clean.", "", result) #keep characters after X
colnames(gcount_filt)<-result #keep just tube number 
```

Select desired metadata columns.    
```{r}
metadata<-metadata%>%
  select(SampleName, Date, Temperature)

metadata$Date<-as.character(metadata$Date)
metadata$Temperature<-as.character(metadata$Temperature)
```

Make sure the metadata and the columns in the gene count matrix are all the same.  
```{r}
metadata$SampleName
colnames(gcount_filt)
```

Reorder them automatically.  

```{r}
list<-colnames(gcount_filt)
list<-as.factor(list)

metadata$SampleName<-as.factor(metadata$SampleName)

# Re-order the levels
metadata$SampleName <- factor(as.character(metadata$SampleName), levels=list)
# Re-order the data.frame
metadata_ordered <- metadata[order(metadata$SampleName),]
metadata_ordered$SampleName

head(metadata_ordered)

metadata_ordered$SampleName
colnames(gcount_filt)
```

These match!    

# **Construct DESeq2 data set by temperature**  

Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at temperature to test for any differences in gene expression across temperatures  
```{r}
#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = gcount_filt,
                              colData = metadata_ordered,
                              design = ~Temperature)
```

First we are going to log-transform the data using a variance stabilizing transforamtion (VST). This is only for visualization purposes. Essentially, this is roughly similar to putting the data on the log2 scale. It will deal with the sampling variability of low counts by calculating within-group variability (if blind=FALSE). Importantly, it does not use the design to remove variation in the data, and so can be used to examine if there may be any variability do to technical factors such as extraction batch effects.

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4.

Chunk should return TRUE if <4.  
```{r}
SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than 4 for us to use vst
print(sizeFactors(SF.gdds)) #View size factors

all(sizeFactors(SF.gdds)) < 4
```

All size factors are less than 4, so we can use VST transformation.  

```{r}
gvst <- vst(gdds, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst), 3) #view transformed gene count data for the first three genes in the dataset.  
```

# **Examine PCA and sample distances of all genes**  

Plot a PCA of samples by temperature.      

```{r}
gPCAdata <- plotPCA(gvst, intgroup = c("Temperature", "Date"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) #plot PCA of samples with all data

allgenesfilt_PCA <- ggplot(gPCAdata, aes(PC1, PC2, color=Temperature, shape=Date)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed()+
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines 
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) # + #Set the plot background
allgenesfilt_PCA
ggsave("Pacu2021/Figures/TagSeq/allgenesfilt-PCA_Temperature.pdf", allgenesfilt_PCA, width=11, height=8)
```

Plot a PCA of samples by day      

```{r}
gPCAdata <- plotPCA(gvst, intgroup = c("Date"), returnData=TRUE)
percentVar <- round(100*attr(gPCAdata, "percentVar")) #plot PCA of samples with all data

allgenesfilt_PCA <- ggplot(gPCAdata, aes(PC1, PC2, color=Date)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed()+
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines 
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) # + #Set the plot background
allgenesfilt_PCA
ggsave("Pacu2021/Figures/TagSeq/allgenesfilt-PCA_Date.pdf", allgenesfilt_PCA, width=11, height=8)
```



NEED TO RUN THESE ANALYSES WHERE WE PARTITION BY DAY - RUN 31 vs 27 on 05 and 06 and run 34 vs 27 on 07


# **Run DEG analysis of DEGs for 27C vs 31C**  

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG <- DESeq(gdds) #run differential expression test by group using the Wald model
```

Explore significant p-values for comparison between bleached and non-bleached corals and generate .csv file.    
```{r, message = FALSE}
DEG.results <- results(DEG, contrast= c("Temperature", "27","31"))
head(DEG.results)
sum(DEG.results$padj < 0.05, na.rm=TRUE)

#there are 14 differentially expressed genes 
DEGs <- as.data.frame(subset(DEG.results, padj<0.05))
DEGs$contrast <- as.factor(c("27_vs_31"))
DEGs$gene_id  <- rownames(DEGs)
rownames(DEGs) <- NULL

write.csv(DEGs, "Pacu2021/Output/TagSeq/DEGs_27_31.csv")
```
There are 14 differentially expressed genes between 27 and 31.    

# **Run DEG analysis of DEGs for 27C vs 34C**  

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG <- DESeq(gdds) #run differential expression test by group using the Wald model
```

Explore significant p-values for comparison between bleached and non-bleached corals and generate .csv file.    
```{r, message = FALSE}
DEG.results <- results(DEG, contrast= c("Temperature", "27","34"))
head(DEG.results)
sum(DEG.results$padj < 0.05, na.rm=TRUE)

#there are X differentially expressed genes 
DEGs <- as.data.frame(subset(DEG.results, padj<0.05))
DEGs$contrast <- as.factor(c("27_vs_34"))
DEGs$gene_id  <- rownames(DEGs)
rownames(DEGs) <- NULL

write.csv(DEGs, "Pacu2021/Output/TagSeq/DEGs_27_34.csv")
```
There are 562 differentially expressed genes between 27 and 34. 

# **Plot a heatmap of sample to sample distances**    

```{r}
gsampleDists <- dist(t(assay(gvst))) #calculate distance matix
gsampleDistMatrix <- as.matrix(gsampleDists) #distance matrix
rownames(gsampleDistMatrix) <- colnames(gvst) #assign row names
colnames(gsampleDistMatrix) <- NULL #assign col names
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   pdf(filename, width=width, height=height)
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}

pht<-pheatmap(gsampleDistMatrix, #plot matrix
         clustering_distance_rows=gsampleDists, #cluster rows
         clustering_distance_cols=gsampleDists, #cluster columns
         col=colors) #set colors

save_pheatmap_pdf(pht, "Pacu2021/Figures/TagSeq/pheatmap.pdf")

```


